core/src/main/scala/org/apache/spark/api/python/PythonGatewayServer.scala

```diff
@@ -43,12 +43,17 @@ private[spark] object PythonGatewayServer extends Logging {
     // with the same secret, in case the app needs callbacks from the JVM to the underlying
     // python processes.
     val localhost = InetAddress.getLoopbackAddress()
-    val gatewayServer: GatewayServer = new GatewayServer.GatewayServerBuilder()
-      .authToken(secret)
+    val builder = new GatewayServer.GatewayServerBuilder()
       .javaPort(0)
       .javaAddress(localhost)
       .callbackClient(GatewayServer.DEFAULT_PYTHON_PORT, localhost, secret)
-      .build()
+    if (sys.env.getOrElse("_PYSPARK_INSECURE_GATEWAY", "0") != "1") {
+      builder.authToken(secret)
+    } else {
+      assert(sys.env.getOrElse("SPARK_TESTING", "0") == "1",
+        "Creating insecure Java gateways only allowed for testing")
```

Comment:
seems a bit weird to say only allowed for testing when isn't this being specifically added for Zeppelin for compatibility.  I would put a bigger warning here about running insecure might leave you vulnerable to attack

Reply:
sorry see you have a stricter warning below

Reply:
this warning is different than the others -- for zeppelin etc., the gateway is created *outside* of spark, and brought in.  Here, we're creating an insecure gateway ourselves, which we should only do for testing.  We dont' want spark itself to ever create an insecure gateway